package/ota_daemon/src/ota_daemon.c

```diff
@@ -0,0 +1,529 @@
+  }
+
+  return ret;
+}
+
+/* TODO: check that token includes only digits */
+#define CHECK_TOKEN(version) do {                                   \
+  if (token) {                                                      \
+    version = atoi(token);                                          \
+  } else {                                                          \
+    piksi_log(LOG_ERR, "Invalid version string: %s", str);          \
+    free(digits);                                                   \
+    return 1;                                                       \
+  }                                                                 \
+} while (0)
+
+static int ota_parse_version_str(const char *str, ota_version_t *ver) {
+
+  size_t len = strlen(str);
+  size_t idx = 0;
+
+  /* Find first digit in the str */
+  while (!isdigit(str[idx])) {
+    idx++;
+
+    if (idx >= len) {
+      piksi_log(LOG_ERR, "Invalid version string: %s", str);
+      return 1;
+    }
+  }
+
+  char *digits = strdup(str + idx);
+  char *rest = digits;
+  char *token = NULL;
+
+  token = strtok_r(rest, ".", &rest);
+  CHECK_TOKEN(ver->major);
+
+  token = strtok_r(NULL, ".", &rest);
+  CHECK_TOKEN(ver->minor);
+
+  token = strtok_r(NULL, ".", &rest);
+  CHECK_TOKEN(ver->build);
+
+  free(digits);
+
+  return 0;
+}
+
+/* 
+ * return:
+ *   > 0 if a newer
+ *   = 0 if a and b equal
+ *   < 0 if b newer
+ */
+static int ota_cmp_version(ota_version_t *a, ota_version_t *b) {
+  if (a->major != b->major) {
+    return a->major - b->major;
+  }
+
+  if (a->minor != b->minor) {
+    return a->minor - b->minor;
+  }
+
+  return a->build - b->build;
+}
+
+/* 
+ * return:
+ *   > 0 if offered version is newer
+ *   = 0 if offered and current versions are equal
+ *   < 0 if current version is newer or error
+ */
+static int ota_version_check(const char *offered) {
+  char current[64];
+  device_fw_version_get(current, sizeof(current));
+
+  ota_version_t offered_parsed = {0};
+  if (ota_parse_version_str(offered, &offered_parsed)) {
+    return -1;
+  }
+
+  ota_version_t current_parsed = {0};
+  if (ota_parse_version_str(current, &current_parsed)) {
+    return -1;
+  }
+
+  int ret = ota_cmp_version(&offered_parsed, &current_parsed);
+
+  if (ret > 0) {
+    piksi_log(LOG_INFO, "OTA new version available: %s", offered);
+  }
+
+  return ret;
+}
+
+static int ota_parse_response(ota_resp_t *parsed_resp) {
+  /* Read and mmap the json file */
+  int fd = open(OTA_RESPONSE_FILE_PATH, O_RDONLY);
+  if (fd == -1) {
+    return 1;
+  }
+
+  struct stat sb = {0};
+
+  if (fstat(fd, &sb) == -1) {
+    close(fd);
+    return 1;
+  }
+
+  void *fmap = mmap(NULL, sb.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+  close(fd);
+  if (fmap == MAP_FAILED) {
+    return 1;
+  }
+
+  /* Parse json into a json_object struct */
+  struct json_object *fjson = json_tokener_parse(fmap);
+
+  struct json_object *json_url = json_object_object_get(fjson, "url");
+  strncpy(parsed_resp->url,
```

Comment:
Switch to snprintf and log a warning if we overflow these buffers?

Reply:
Replaced `CHECKED_SPRINTF` macro with `snprintf_assert` and `snprintf_warn` wrapper functions.

Not sure if I followed to the fullest with the latter part of the comment so I didn't act on in. Please elaborate if further actions are desired.

Reply:
We can leave this out of this PR if you don't feel it's appropriate but I was thinking it might be a good idea to have some defensive code in the download of resources from the remote system.  Particularly in `network_download_write` we can have 2 new parameters has part of the `ctx` struct a counter of the current amount of data downloaded and a limit.  For JSON blob downloads, this should be limited to 4k, for firmware download it should limited to 30mb.

Reply:
The fixes to `CHECKED_SPRINTF` look good though, thanks.

Reply:
I think I'll leave that to followup-PR. There's already quite a few diversions in this PR already and I'd like to get this wrapped inside this sprint if possible.

Reply:
Created an issue https://swift-nav.atlassian.net/projects/DEVC/issues/DEVC-1141 and start looking into this.